WooCommerce/Classes/POS/Utils/POSProductProvider.swift

```diff
@@ -0,0 +1,90 @@
+    private var siteID: Int64
+    private var currencySettings: CurrencySettings
+
+    init() {
+        self.storageManager = ServiceLocator.storageManager
+        self.siteID = ServiceLocator.stores.sessionManager.defaultSite?.siteID ?? 0
+        self.currencySettings = ServiceLocator.currencySettings
+    }
+
+    private lazy var productsResultsController: ResultsController<StorageProduct> = {
+        let predicate = NSPredicate(format: "siteID == %lld", siteID)
+        let descriptor = NSSortDescriptor(key: "name", ascending: true, selector: #selector(NSString.caseInsensitiveCompare))
+        let resultsController = ResultsController<StorageProduct>(storageManager: storageManager,
+                                                                  matching: predicate,
+                                                                  sortedBy: [descriptor])
+        return resultsController
+    }()
+
+    /// Provides a`[POSProduct]`array by mapping  simple, purchasable-only Products from storage
+    ///
+    func providePointOfSaleProducts() -> [POSProduct] {
+        var loadedProducts: [Product] = []
+        var pointOfSaleProducts: [POSProduct] = []
+
+        // 1. Fetch products from storage, and filter them by `purchasable` and `simple`
+        do {
+            try productsResultsController.performFetch()
+            if productsResultsController.fetchedObjects.isEmpty {
+                // TODO: Handle case for empty product list, or not empty but no eligible products
+                // https://github.com/woocommerce/woocommerce-ios/issues/12815
+                // https://github.com/woocommerce/woocommerce-ios/issues/12816
+                DDLogWarn("No products eligible for POS, or empty storage.")
+            } else {
+                // Ideally we should handle the filtering through a policy that can be easily modified,
+                // rather than having this declared implicitely:
+                loadedProducts = productsResultsController.fetchedObjects.filter { $0.productType == .simple && $0.purchasable }
+            }
+        } catch {
+            // TODO: Handle case for error when fetching products
+            // https://github.com/woocommerce/woocommerce-ios/issues/12846
+            DDLogError("Error fetching products from storage")
+        }
+
+        // 2. Map result to POSProduct and populate the output
+        for product in loadedProducts {
+            let posProduct = POSProduct(itemID: UUID(),
+                                        productID: product.productID,
+                                        name: product.name,
+                                        price: product.price,
+                                        stockQuantity: product.stockQuantity?.intValue ?? 0, // TODO: Remove stock usage
+                                        currencySettings: currencySettings)
+            pointOfSaleProducts.append(posProduct)
+        }
+        return pointOfSaleProducts
```

Comment:
nit: how about mapping `loadedProducts` to return? like:

 ```suggestion
        return loadedProducts.map { product in
            POSProduct(itemID: UUID(),
                                        productID: product.productID,
                                        name: product.name,
                                        price: product.price,
                                        stockQuantity: product.stockQuantity?.intValue ?? 0, // TODO: Remove stock usage
                                        currencySettings: currencySettings)
        }
```

Reply:
Sounds good! It also reduces the need to declare a temporary empty array for POSProducts. Updated on https://github.com/woocommerce/woocommerce-ios/pull/12845/commits/7e0d86a1cc0d9f0967bc2589ff31def799a6b76d